openadapt/models_tuning/fine_tune_models.py

```diff
@@ -0,0 +1,207 @@
+    x2, y2 = point2
+    distance = math.sqrt((x2 - x1) ** 2 + (y2 - y1) ** 2)
+    return distance
+
+
+def _calculate_similarity_per_action(ref_action, prediction_action):
+    # verify if the 2 actions contains the same keys
+    is_same_action_type = ref_action.get("name") == prediction_action.get("name")
+    if not is_same_action_type:
+        return 0
+    if ref_action.get("name") in ("press", "release"):
+        is_same_key = ref_action.get("canonical_key_char") == prediction_action.get(
+            "canonical_key_char"
+        )
+        if not is_same_key:
+            return 0
+        return 1
+    else:
+        # calculate euclidean distance between the mouse positions
+        ref_mouse_pos = (ref_action.get("mouse_x"), ref_action.get("mouse_y"))
+        prediction_mouse_pos = (
+            prediction_action.get("mouse_x"),
+            prediction_action.get("mouse_y"),
+        )
+        distance = _euclidean_distance(ref_mouse_pos, prediction_mouse_pos)
+        # normalize the distance between 0 and 1
+        normalized_distance = distance / MAX_SCREEN_SIZE[0]
+        # the smaller the distance, the better the score
+        return 1 - normalized_distance
+
+
+def _generate_prediction_dataset(model, tokenizer, dataset):
+    """
+    Generate a prediction dataset based on the given dataset.
+    """
+    prediction_dataset = []
+    count = 0
+    for entry in dataset:
+        if count > 10:
+            return prediction_dataset
+        reference_window_dict = entry["reference_window_dict"]
+        reference_action_dicts = entry["reference_action_dicts"]
+        active_window_dict = entry["active_window_dict"]
+        reference_window_dict["state"].pop("data")
+
+        prompt = (
+            f"{reference_window_dict=}\n{reference_action_dicts=}\n{active_window_dict=}\nProvide"
+            " valid Python3 code containing the action dicts by completing the"
+            " following, and nothing else:\nactive_action_dicts="
+        )
+        if len(prompt) > MAX_INPUT_SIZE:
+            logger.warning(f"Truncating from {len(prompt) =} to {MAX_INPUT_SIZE=}")
+            prompt = prompt[-MAX_INPUT_SIZE:]
+            logger.warning(f"Truncated {len(prompt)=}")
+        input_tokens = tokenizer(prompt, return_tensors="pt")
+        pad_token_id = tokenizer.eos_token_id
+        attention_mask = input_tokens["attention_mask"]
+        output_tokens = model.generate(
+            input_ids=input_tokens["input_ids"],
+            attention_mask=attention_mask,
+            pad_token_id=pad_token_id,
+            max_length=1000,
+            num_return_sequences=1,
+        )
+        N = input_tokens["input_ids"].shape[-1]
+        completion = tokenizer.decode(
+            output_tokens[:, N:][0],
+            clean_up_tokenization_spaces=True,
+        )
+        active_action_dicts = get_action_dict_from_completion(completion)
+        logger.debug(f"active_action_dicts=\n{pformat(active_action_dicts)}")
```

Comment:
@abrichr May be am I wrong here, but using local `gpt-2` model from `transformers` package, I do not get a lot of correct prediction. Must of the prediction given by `gpt-2` model cannot be parsed to correct `action_event`.

Is there any special tricks that I miss ?

Reply:
If this is truly the case, then either my understand of how to build the `prompt` from Event is incorrect, or we need to re-consider the way a prompt is built.

Otherwise, the whole point of improving the model becomes less critical as may be the issue is not the model, but our prompt building itself.

WDYT ? @abrichr

Reply:
@LaPetiteSouris that you for the information regarding GPT-2! We have been testing with GPT3.5-turbo and GPT-4.

For enforcing output structure, we have looked at implementing [guardrails](https://github.com/OpenAdaptAI/OpenAdapt/issues/26) and [LMQL/Guidance](https://github.com/OpenAdaptAI/OpenAdapt/issues/227), but so far without much luck. If you could assist in this as well that would be greatly appreciate 🙏 

You are correct that we need to implement evaluation metrics. Related: https://github.com/OpenAdaptAI/OpenAdapt/issues/173 https://github.com/OpenAdaptAI/OpenAdapt/issues/414

Reply:
@abrichr I'll take a look into LMQL Guidance/guardrails topic as soon as I wrap up my work on these model evaluations/tuning topic.